Practice/calender

YudaiNoda3576 opened the pull request:
## 目的
カレンダープログラムの課題を提出すること

## 達成条件
レビュワーからマージ許可を頂くこと

##  仕様
- `m`で月を、`y`で年を指定できる
    - ただし、`y`のみ指定して一年分のカレンダーを表示する機能の実装は不要
- 引数を指定しない場合は、今月・今年のカレンダーが表示される
- macに入っているcalコマンドと同じ見た目になっている
    
   *  ヘッダ部最上部にM月 yyyyの表示
    
    * ヘッダ部に日月火水木金土
    
    * calコマンドで表示される完成物イメージ
![calender_sample](https://user-images.githubusercontent.com/63564761/160066397-6fcb5383-50cc-4933-96ee-343453035712.png)

- 少なくとも1970年から2100年までは正しく表示される

- Rubyの標準ライブラリだけを使うこと

## このプルリクでしないこと
* テストコードの作成

## レビュー希望箇所
* 32/33行目の、今月が何週間あるかを判定するロジックがもっとスマートにできる気がしますが、他に方法がありましたらご教示ください。
* 全体的に手続的な記述方法になっているように感じます。より宣言的に記述する場合、修正箇所・修正案をご教示頂けましたら幸いです。

## 懸念事項
特になし

## 保留事項
* 今日の日付の部分の色が反転する
* どのような引数が与えられようが、cal コマンドと同じ表示結果になる

## 備考

YudaiNoda3576 committed:
feat

現在の年月のカレンダーをコマンドで出力する機能を実装

YudaiNoda3576 committed:
fix

ファイル名の変更、コマンドラインからの入力で引数を受け取るように変更

YudaiNoda3576 committed:
fix

pushされたファイルの削除

YudaiNoda3576 committed:
feat

mycalというコマンドで実行出来るように実装

betachelsea commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,56 @@
+  #nilの場合の対処として、空欄を入れる
+  if week[6] == nil
+    week.map { |e| e ? e : " " }
+  end
```
[MUST] おそらく、最終週について処理しようとされていると思いますが、残念ながら特に機能していないようです。
例えば、2022年3月の最終週は 27(日)〜31(木)なので、

```ruby
week = [27, 28, 29, 30, 31]
```

を処理することになります。その場合、 `week.map` はこの5要素を繰り返して新たな配列をつくるため、`e ? e : " "` の三項演算子の `e` には nil は入ってこないようです。

また、作った後の配列を保存していないようなので、特に利用されている様子もなさそうと思いました 👀 
(転じていえば、無くてもちゃんと動いているということですね！)

この処理はカットしてもらいたいです 🙏  よろしくお願いします！

betachelsea commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,56 @@
+    week.map { |e| e ? e : " " }
+  end
+  #1週分を出力する
+  printf("%3s%3s%3s%3s%3s%3s%3s\n",
+         week[0], week[1], week[2], week[3], week[4], week[5], week[6])
```
[nits] 何回も書いているのをみると短く済ませたくなりますね！
`week.each` などで対処できるような気がしますので、よければトライしてみてほしいです！ 👀

betachelsea commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,56 @@
+
+options = ARGV.getopts("y:", "m:")
+
+year = options['y'].to_i
+month = options['m'].to_i
+
+if year == 0
+  year = today.year
+end
+
+if month == 0
+  month = today.month
+end
```
[nits] rubyは次のように書くことで、「aが nil なら b」、のような代入をすることができます。

```ruby
a = nil
b = 1
c = 2

x = a || b
puts x #=> 1

y = a || b || c
puts y #=> 1  ※より左に書いてあるものが優先
```

この記法を活用し、次のように書くことができそうです 👀 

```ruby
year = (options['y'] || today.year).to_i
```

betachelsea reviewed (commented):
提出ありがとうございます！少しコメントしました 📝 

> 32/33行目の、今月が何週間あるかを判定するロジックがもっとスマートにできる気がしますが、他に方法がありましたらご教示ください。

もっと考えるとすれば、次のように思いました 🤔 

- 毎月の日数は30 or 31で、2月だけ28 or 29日
- 4週間は必ずあるとして ( 28 / 7 = 4 ) 計算を省略し、そこに +1 するかどうかだけを考慮しても良いのでは？

> 全体的に手続的な記述方法になっているように感じます。より宣言的に記述する場合、修正箇所・修正案をご教示頂けましたら幸いです。

現時点ではこの形で問題ないと思います！ まずRubyの書き方に慣れていただいたあと、オブジェクト指向の書き方を実施してもらうプラクティスも後々ありますので、ご安心ください！ 🙏

YudaiNoda3576 committed:
fix

【Must】
最終週について判定する不要なロジックを削除
【nits】
year・monthの代入式の修正
今月が何週間あるかを判定するロジックの修正
週ごとの出力のロジックを修正

betachelsea commented on 02.calendar/cal.rb:
```diff
@@ -0,0 +1,46 @@
+  #1週分の内容weekにセットする
+  week = days_of_month.shift(7)
+  #1週分を出力する
+  week.each_with_index do |w, i|
+    i == 6 ? printf("%3s\n", w) : printf("%3s", w)
+  end
```
[IMO] ハードコードな繰り返し感が減って、すごく良くなりましたね！ありがとうございます！ 🙌 

別解になりますが、 上の行で `shift(7)` しているので、最後に改行すれば良いと考えると、次のようにすることもできそうです。

```ruby
 week.each {|w| printf("%3s", w) }
 print("\n")
```

betachelsea reviewed (approved):
修正ありがとうございます！確認OKです 👍 

ソースコード全体についてですが、かなりコメントを細かく記入されていますね。
最初Rubyに慣れないうちは学習も兼ねてというところかなと思いますが、少しずつ「必要最低限のコメント」に慣れていっていただきたいです！
ソースコード自体から何を作りたかったかが透けて見え、コメントが不要である...という形が理想です。
もちろん、複雑な処理をしているところではコメントが必要なケースもあると思いますが、1行ずつではなく処理の塊( 今後慣れていく「メソッド」など )の単位で書くことが多いかなと思います。

例えば

```ruby
first_day = Date.new(year, month, 1).wday
```

この行は表記から指定のyear, monthの月初(1日)の曜日を変数に代入したい、ということは明らかかなと感じます。
(改めて見ると `first_day` の day は曜日という意味にも日付という意味にも取れるので、あえてRubyのメソッド名を拝借して `first_wday` とかでも良いかなと感じますが... 命名は難しいですね！ 😅  )

今後のプラクティスでは、不安な点もあるかと思いますが、コメントの量を減らしていく方向にぜひシフトしていっていただきたいです。よろしくお願いします〜！ 👍

YudaiNoda3576 committed:
fix

不要なコメントの削除
